### PR TITLE
furureObservation

### LINKED ISSUE #117: Prior history with multiple observation periods
@catalamarti at the moment addPriorHistory will return multiple rows when a person has multiple observation periods which I don´t think is the desired behaviour, right? In addDemographics I made it so that the most recent observation start date would be used - is that the behaviour you´d want?

One thing to note with my implementation though (which you can see in the second example) is that if the first observation period starts after the current index date then a NA will be returned instead of a negative number. Is that ok?

``` r
library(PatientProfiles)
# with multiple observation periods for person id 1
person <- tibble::tibble(
  person_id = c(1,2),
  gender_concept_id = 1,
  year_of_birth = 1980,
  month_of_birth = 01,
  day_of_birth = 01
)
observation_period <- tibble::tibble(
  observation_period_id = c(1,2, 3),
  person_id = c(1,1, 2),
  observation_period_start_date = c(as.Date("2000-01-01"),
                                    as.Date("2010-01-01"),
                                    as.Date("2010-01-01")),
  observation_period_end_date = c(as.Date("2005-01-01"),
                                  as.Date("2015-01-01"),
                                  as.Date("2015-01-01"))
)
cohort1 <- tibble::tibble(
  cohort_definition_id = 1,
  subject_id = c(1,2),
  cohort_start_date = as.Date(c("2012-02-01")),
  cohort_end_date = as.Date(c("2013-02-01"))
)

cdm <- mockPatientProfiles(person = person,
                           observation_period = observation_period,
                           cohort1 = cohort1
)

cdm$cohort1 %>%
  addPriorHistory(cdm,
                  indexDate = "cohort_start_date"
  )
#> # Source:   table<dbplyr_001> [3 x 5]
#> # Database: DuckDB 0.5.0 [unknown@Linux 5.4.0-139-generic:R 4.1.2/:memory:]
#>   cohort_definition_id subject_id cohort_start_date cohort_end_date prior_hist…¹
#>                  <dbl>      <dbl> <date>            <date>                 <dbl>
#> 1                    1          1 2012-02-01        2013-02-01               761
#> 2                    1          2 2012-02-01        2013-02-01               761
#> 3                    1          1 2012-02-01        2013-02-01              4414
#> # … with abbreviated variable name ¹​prior_history

cdm$cohort1 %>%
  addDemographics(cdm,
                  indexDate = "cohort_start_date",
                  age = FALSE,
                  ageGroup = NULL,
                  sex = FALSE,
                  priorHistory = TRUE
  )
#> # Source:   table<dbplyr_002> [2 x 5]
#> # Database: DuckDB 0.5.0 [unknown@Linux 5.4.0-139-generic:R 4.1.2/:memory:]
#>   cohort_definition_id subject_id cohort_start_date cohort_end_date prior_hist…¹
#>                  <dbl>      <dbl> <date>            <date>                 <dbl>
#> 1                    1          1 2012-02-01        2013-02-01               761
#> 2                    1          2 2012-02-01        2013-02-01               761
#> # … with abbreviated variable name ¹​prior_history
```

<sup>Created on 2023-03-30 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

NAs 
``` r
library(PatientProfiles)
# with multiple observation periods for person id 1
person <- tibble::tibble(
  person_id = c(1,2),
  gender_concept_id = 1,
  year_of_birth = 1980,
  month_of_birth = 01,
  day_of_birth = 01
)
observation_period <- tibble::tibble(
  observation_period_id = c(1,2, 3),
  person_id = c(1,1, 2),
  observation_period_start_date = c(as.Date("2000-01-01"),
                                    as.Date("2010-01-01"),
                                    as.Date("2010-01-01")),
  observation_period_end_date = c(as.Date("2005-01-01"),
                                  as.Date("2015-01-01"),
                                  as.Date("2015-01-01"))
)
cohort1 <- tibble::tibble(
  cohort_definition_id = 1,
  subject_id = c(1,2),
  cohort_start_date = as.Date(c("2006-02-01")),
  cohort_end_date = as.Date(c("2006-02-01"))
)

cdm <- mockPatientProfiles(person = person,
                           observation_period = observation_period,
                           cohort1 = cohort1
)

cdm$cohort1 %>%
  addDemographics(cdm,
                  indexDate = "cohort_start_date",
                  age = FALSE,
                  ageGroup = NULL,
                  sex = FALSE,
                  priorHistory = TRUE
  )
#> # Source:   table<dbplyr_001> [2 x 5]
#> # Database: DuckDB 0.5.0 [unknown@Linux 5.4.0-139-generic:R 4.1.2/:memory:]
#>   cohort_definition_id subject_id cohort_start_date cohort_end_date prior_hist…¹
#>                  <dbl>      <dbl> <date>            <date>                 <dbl>
#> 1                    1          1 2006-02-01        2006-02-01              2223
#> 2                    1          2 2006-02-01        2006-02-01                NA
#> # … with abbreviated variable name ¹​prior_history
```

<sup>Created on 2023-03-30 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

### PR DESCRIPTION
**Pull request by edward-burn**
- add furureObservation to addDemographics #120
- create addFurureObservation function
- abstract out some of single add* functions to use in both specific functions and addDemographics
- fix #117

### CODE REVIEW

File: R/addDemographics.R [hunk 1]
```diff
@@ -133,7 +138,9 @@ addDemographics <- function(x,
                                          !!rlang::sym(indexDate)) %>%
                          dplyr::group_by(dplyr::across(dplyr::all_of(c("person_id", indexDate)))) %>%
                          dplyr::summarise(observation_period_start_date =
```
Comment (catalamarti) [inline R/addDemographics.R:140]: is this summarise necessary?

Comment (edward-burn) [reply R/addDemographics.R]: True! Not any more as we´re using the current observation period - have dropped this in the latest commit

**Comment by catalamarti** [comment]
Hi @edward-burn I added a failing test

**Comment by edward-burn** [comment]
> Hi @edward-burn I added a failing test

@catalamarti the latest commit should have fixed this case of multiple rows and multiple observation periods. I think the way I´ve done it could likely be improved in a refactoring though ....

I only fixed this for addDemographics though (because the others would also be fixed if we do https://github.com/oxford-pharmacoepi/PatientProfiles/issues/129)

**Comment by catalamarti** [comment]
Hi @edward-burn can you take a look if my last commit makes sense? Then I will merge

**Review by catalamarti — APPROVED** [review]

**Pull request merged**